types/react/test/hooks.tsx

```diff
@@ -133,22 +144,23 @@ function useEveryHook(ref: React.Ref<{ id: number }>|undefined): () => boolean {
     React.useRef(undefined);
 
     // |null convenience overload
-    // it should _not_ be mutable if the generic argument doesn't include null
-    // $ExpectType RefObject<number>
+    // $ExpectType MutableRefObject<number | null>
     React.useRef<number>(null);
-    // but it should be mutable if it does (i.e. is not the convenience overload)
     // $ExpectType MutableRefObject<number | null>
     React.useRef<number | null>(null);
+    // $ExpectType MutableRefObject<number | null | undefined>
+    React.useRef<number | null | undefined>(null);
 
     // |undefined convenience overload
     // with no contextual type or generic argument it should default to undefined only (not {} or unknown!)
     // $ExpectType MutableRefObject<undefined>
     React.useRef();
     // $ExpectType MutableRefObject<number | undefined>
     React.useRef<number>();
-    // don't just accept a potential undefined if there is a generic argument
-    // $ExpectError
+    // $ExpectType MutableRefObject<number | undefined>
     React.useRef<number>(undefined);
```

Comment:
Are we sure this is what we want? For literal values I can see the point. But what if we have e.g. `React.useRef<number>(props.someNullableNumber)`? I guess type-checking still works when accessing this ref but it might be passed to something that's not type-safe (e.g. `doSomethingWithANonNullableNumber(ref.current)`) which means the code is no longer type-safe. Before, we could at least have a stop-gap when instantiating the ref.

What happens for `React.useRef<number>('not a number')`? I would expect an error instead of `MutableRefObject<number | string>`. It doesn't look like we have a test for that.

Reply:
> Are we sure this is what we want? For literal values I can see the point. But what if we have e.g. React.useRef<number>(props.someNullableNumber)?

I'd be open to letting `useRef<number>(foo as number | null)` give a type error.

In fact, the thing I'm most interested in is `useRef<number>()` returning `MutableRefObject<number | undefined>`.

> I guess type-checking still works when accessing this ref but it might be passed to something that's not type-safe (e.g. doSomethingWithANonNullableNumber(ref.current)) which means the code is no longer type-safe. Before, we could at least have a stop-gap when instantiating the ref.

In this case `ref.current` will be typed as `number | null` so the type checking would work as expected.

> What happens for React.useRef<number>('not a number')? I would expect an error instead of MutableRefObject<number | string>. It doesn't look like we have a test for that.

Pushed a commit that adds these tests:

<img width="318" alt="Screenshot 2021-11-30 at 18 35 15" src="https://user-images.githubusercontent.com/189580/144098517-3af8130c-b5d1-47f2-b0f0-0e40a282b491.png">

Reply:
> In fact, the thing I'm most interested in is useRef<number>() returning MutableRefObject<number | undefined>.

I'm in favor of this behavior considering this is equivalent to a literal `undefined`. There's still room for a mistake but that mistake is just a few code columns away so I think that we don't necessarily need to catch it.

 > In this case ref.current will be typed as number | null so the type checking would work as expected.

Intended by you maybe but the resulting behavior is not necessarily expected. Consider you pass this `ref.current` to a function that's loosely typed e.g. `handleRef(value: any)`. Before, you could be sure that `useRef` prevents these mistakes. Your approach assumes that the whole codebase is type-checked which is rarely the case (especially with library types).

Reply:
Maybe this wasn't very clear: 
To me `const numberRef = useRef<number>()` might be a mistake (forgot to pass the initial value). That mistake is caught when trying to access `numberRef.current` and expecting a number. However, when `numberRef` is passed to a part of the codebase that is untyped e.g. `adder(numberRef.current, 2)` where `adder(a: any, b: any): number` then we'll never catch this mistake at the type level.

The unfortunate part is that we'll never know which scenario is more likely. But considering you can already get the desired behavior with `useRef<number | undefined>` I'm inclined to not merge this proposal.

My underlying principal is that I try to publish types that are strict by default. The library user can always loosen them up but rarely make them stricter. That's not matching the TypeScript mentality (pragmatically typed language) but I don't find this proposal convincing with regards to pragmatism.

Though the real problem is that we're just two people discussing this. I wouldn't block this proposal if another maintainer would vote for merging this proposal.